graphql_compiler/compiler/ir_lowering_common.py

```diff
@@ -159,3 +161,46 @@ def visitor_fn(expression):
         new_ir_blocks.append(new_block)
 
     return new_ir_blocks
+
+
+def extract_folds_from_ir_blocks(ir_blocks):
+    """Extract all @fold data from the IR blocks, and cut the folded IR blocks out of the IR.
+
+    Args:
+        ir_blocks: list of IR blocks to extract fold data from
+
+    Returns:
+        tuple (folds, remaining_ir_blocks):
+        - folds: dict of FoldScopeLocation -> list of IR blocks corresponding to that @fold scope.
+                 The list does not contain Fold or Unfold blocks.
+        - remaining_ir_blocks: list of IR blocks that were not part of a Fold-Unfold section.
+    """
+    folds = dict()
+    remaining_ir_blocks = []
+    current_folded_blocks = []
+    in_fold_location = None
+
+    for block in ir_blocks:
+        if isinstance(block, Fold):
+            if in_fold_location is not None:
```

Comment:
I can only assume that the first `ir_block` can never be a fold ?

Reply:
The first block cannot be a `Fold`, though this is not what the assertion checks. The assertion checks that you don't have a `Fold` within a `Fold`.